core/state-machine/src/lib.rs

```diff
@@ -167,48 +163,52 @@ fn always_untrusted_wasm<E, R: Decode>() -> ExecutionManager<DefaultHandler<R, E
 }
 
 /// The substrate state machine.
-pub struct StateMachine<'a, B, H, N, T, O, Exec> {
-	backend: B,
-	changes_trie_storage: Option<&'a T>,
-	offchain_ext: Option<&'a mut O>,
-	overlay: &'a mut OverlayedChanges,
+pub struct StateMachine<'a, B, H, N, T, Exec> where H: Hasher<Out=H256>, B: Backend<H> {
+	backend: &'a B,
 	exec: &'a Exec,
 	method: &'a str,
 	call_data: &'a [u8],
-	keystore: Option<BareCryptoStorePtr>,
-	_hasher: PhantomData<(H, N)>,
+	ext: Ext<'a, H, N, B, T>,
 }
 
-impl<'a, B, H, N, T, O, Exec> StateMachine<'a, B, H, N, T, O, Exec> where
-	H: Hasher,
-	Exec: CodeExecutor<H>,
+impl<'a, B, H, N, T, Exec> StateMachine<'a, B, H, N, T, Exec> where
+	H: Hasher<Out=H256>,
+	Exec: CodeExecutor,
 	B: Backend<H>,
 	T: ChangesTrieStorage<H, N>,
-	O: offchain::Externalities,
-	H::Out: Ord + 'static,
 	N: crate::changes_trie::BlockNumber,
 {
 	/// Creates new substrate state machine.
 	pub fn new(
-		backend: B,
+		backend: &'a B,
 		changes_trie_storage: Option<&'a T>,
-		offchain_ext: Option<&'a mut O>,
+		offchain_ext: Option<OffchainExt>,
 		overlay: &'a mut OverlayedChanges,
 		exec: &'a Exec,
 		method: &'a str,
 		call_data: &'a [u8],
 		keystore: Option<BareCryptoStorePtr>,
 	) -> Self {
-		Self {
+		let mut ext = ext::Ext::new(
+			overlay,
 			backend,
 			changes_trie_storage,
-			offchain_ext,
-			overlay,
+		);
+
+		if let Some(keystore) = keystore {
+			ext.register_extension(keystore);
```

Comment:
Hmm, I see that persistent `Ext` is now needed to prevent the extensions from being dropped? Maybe we can extract the `ExtensionStore` then and just pass it to externalities?